src/transaction.rs

```diff
@@ -137,3 +151,139 @@ pub fn remove_spent_notes(
+    let mut builder = Builder::new(network, block_height);
+
+    let fee = 2365000u64;
+
+    let mut total = 0;
+    let mut nullifiers = vec![];
+    for (note, witness) in notes {
+        let witness = Cursor::new(hex::decode(witness)?);
+        let witness = IncrementalWitness::<Node>::read(witness)?;
+        builder
+            .add_sapling_spend(
+                extsk.clone(),
+                *note.recipient().diversifier(),
+                note.clone(),
+                witness.path().ok_or("Commitment Tree is empty")?,
+            )
+            .map_err(|_| "Failed to add sapling spend")?;
+        let nullifier = note.nf(
+            &extsk
+                .to_diversifiable_full_viewing_key()
+                .to_nk(Scope::Internal),
+            witness.position() as u64,
+        );
+        nullifiers.push(hex::encode(nullifier.to_vec()));
+        total += note.value().inner();
+        if total >= amount + fee {
+            break;
+        }
+    }
+
+    if total < amount + fee {
+        Err("Not enough balance")?;
+    }
+    let change = Amount::from_u64(total - amount - fee).map_err(|_| "Invalid change")?;
+    let amount = Amount::from_u64(amount).map_err(|_| "Invalid amount")?;
+    let change_address =
+        decode_payment_address(network.hrp_sapling_payment_address(), change_address)
+            .map_err(|_| "Failed to decode change address")?;
+    if to_address.starts_with(network.hrp_sapling_payment_address()) {
+        let to_address = decode_payment_address(network.hrp_sapling_payment_address(), to_address)
+            .map_err(|_| "Failed to decode sending address")?;
+        builder
+            .add_sapling_output(None, to_address, amount, MemoBytes::empty())
+            .map_err(|_| "Failed to add output")?;
+    } else {
+        let to_address = decode_transparent_address(
+            &network.b58_pubkey_address_prefix(),
+            &network.b58_script_address_prefix(),
+            to_address,
+        )?
+        .ok_or("Failed to decode transparent address")?;
+        builder
+            .add_transparent_output(&to_address, amount)
+            .map_err(|_| "Failed to add output")?;
+    }
+
+    builder
+        .add_sapling_output(None, change_address, change, MemoBytes::empty())
+        .map_err(|_| "Failed to add change")?;
+
+    #[cfg(not(test))]
+    return {
+        let (tx, _metadata) = builder.build(
+            &*PROVER,
+            &FeeRule::non_standard(Amount::from_u64(2365000).map_err(|_| "Invalid fee")?),
+        )?;
+
+        let mut tx_hex = vec![];
+        tx.write(&mut tx_hex)?;
+
+        Ok(JSTransaction {
+            txid: tx.txid().to_string(),
+            txhex: hex::encode(tx_hex),
+            nullifiers,
+        })
+    };
+    #[cfg(test)]
+    {
+        // At this point we would use .mock_build()
+        // However it returns an error for some reason
+        // So let's just return the nullifiers and test those
+        Ok(JSTransaction {
```

Comment:
this test is a bit weak but if we cannot do more it's ok

Reply:
We can't use the actual parameters on the test, so it's our only option